# Lab | Data Aggregation and Filtering

In this challenge, we will continue to work with customer data from an insurance company. We will use the dataset called marketing_customer_analysis.csv, which can be found at the following link:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis.csv

This dataset contains information such as customer demographics, policy details, vehicle information, and the customer's response to the last marketing campaign. Our goal is to explore and analyze this data by first performing data cleaning, formatting, and structuring.

1. Create a new DataFrame that only includes customers who have a total_claim_amount greater than $1,000 and have a response of "Yes" to the last marketing campaign.

In [ ]:
import pandas as pd
df=pd.read_csv('https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis.csv')
(df.head())
new_df=df[(df["Total Claim Amount"] > 1000) & (df["Response"] == "Yes")]
new_df


Unnamed: 0 Customer       State  Customer Lifetime Value Response  \
0           0  DK49336     Arizona              4809.216960       No   
1           1  KX64629  California              2228.525238       No   
2           2  LZ68649  Washington             14947.917300       No   
3           3  XL78013      Oregon             22332.439460      Yes   
4           4  QA50777      Oregon              9025.067525       No   

   Coverage Education Effective To Date EmploymentStatus Gender  ...  \
0     Basic   College           2/18/11         Employed      M  ...   
1     Basic   College           1/18/11       Unemployed      F  ...   
2     Basic  Bachelor           2/10/11         Employed      M  ...   
3  Extended   College           1/11/11         Employed      M  ...   
4   Premium  Bachelor           1/17/11    Medical Leave      F  ...   

   Number of Open Complaints Number of Policies     Policy Type        Policy  \
0                        0.0                  9  Corporate Auto  Corporate L3   
1                        0.0                  1   Personal Auto   Personal L3   
2                        0.0                  2   Personal Auto   Personal L3   
3                        0.0                  2  Corporate Auto  Corporate L3   
4                        NaN                  7   Personal Auto   Personal L2   

   Renew Offer Type  Sales Channel  Total Claim Amount  Vehicle Class  \
0            Offer3          Agent          292.800000  Four-Door Car   
1            Offer4    Call Center          744.924331  Four-Door Car   
2            Offer3    Call Center          480.000000            SUV   
3            Offer2         Branch          484.013411  Four-Door Car   
4            Offer1         Branch          707.925645  Four-Door Car   

  Vehicle Size Vehicle Type  
0      Medsize          NaN  
1      Medsize          NaN  
2      Medsize            A  
3      Medsize            A  
4      Medsize          NaN  

[5 rows x 26 columns]

2. Using the original Dataframe, analyze the average total_claim_amount by each policy type and gender for customers who have responded "Yes" to the last marketing campaign. Write your conclusions.

In [10]:
# Group by policy type and gender, filtering for customers who responded "Yes"
avg_claim_amount = df[df["Response"] == "Yes"].groupby(["Policy Type", "Gender"])["Total Claim Amount"].mean()
print(avg_claim_amount)

Policy Type     Gender
Corporate Auto  F         433.738499
                M         408.582459
Personal Auto   F         452.965929
                M         457.010178
Special Auto    F         453.280164
                M         429.527942
Name: Total Claim Amount, dtype: float64


- Corporate Auto Policies: Female customers who responded to the marketing campaign had a slightly higher average claim amount compared to male customers. This suggests that women might have higher claim costs in this category.

- Personal Auto Policies: The claim amounts are very similar between genders, with men ($457.01) having a slightly higher average than women ($452.97). This indicates that gender does not play a significant role in determining claim amounts for personal auto policies.

- Special Auto Policies: Women again have a slightly higher average claim amount compared to men, showing a small difference in claim behavior.


3. Analyze the total number of customers who have policies in each state, and then filter the results to only include states where there are more than 500 customers.

In [45]:

# Count the number of customers per state
customers_per_state = df.groupby("State")["Customer"].count()
print(customers_per_state)
# Filter states with more than 500 customers
filtered_states = customers_per_state[customers_per_state > 500]
print('States with more than 500 customers')
print(filtered_states)

State
Arizona       1937
California    3552
Nevada         993
Oregon        2909
Washington     888
Name: Customer, dtype: int64
States with more than 500 customers
State
Arizona       1937
California    3552
Nevada         993
Oregon        2909
Washington     888
Name: Customer, dtype: int64


4. Find the maximum, minimum, and median customer lifetime value by education level and gender. Write your conclusions.

In [23]:
CLV_STATS = df.groupby(["Education","Gender"])["Customer Lifetime Value"].agg(["max", "min", "median"])

CLV_STATS 

max          min       median
Education            Gender                                       
Bachelor             F       73225.95652  1904.000852  5640.505303
                     M       67907.27050  1898.007675  5548.031892
College              F       61850.18803  1898.683686  5623.611187
                     M       61134.68307  1918.119700  6005.847375
Doctor               F       44856.11397  2395.570000  5332.462694
                     M       32677.34284  2267.604038  5577.669457
High School or Below F       55277.44589  2144.921535  6039.553187
                     M       83325.38119  1940.981221  6286.731006
Master               F       51016.06704  2417.777032  5729.855012
                     M       50568.25912  2272.307310  5579.099207

## Bonus

5. The marketing team wants to analyze the number of policies sold by state and month. Present the data in a table where the months are arranged as columns and the states are arranged as rows.

In [ ]:
df["Month"] = pd.to_datetime(df["Effective To Date"], format="%m/%d/%y").dt.month# Convert 'Effective_To_Date' to datetime format to extract month

policies_by_state_month = df.groupby(["State", "Month"])["Policy"].count().reset_index()# Group by state and month to count policies sold

table = policies_by_state_month.pivot(index="State", columns="Month", values="Policy")# Pivot to format the table

table



State  Month  Policy
0     Arizona      1    1008
1     Arizona      2     929
2  California      1    1918
3  California      2    1634
4      Nevada      1     551
5      Nevada      2     442
6      Oregon      1    1565
7      Oregon      2    1344
8  Washington      1     463
9  Washington      2     425

6.  Display a new DataFrame that contains the number of policies sold by month, by state, for the top 3 states with the highest number of policies sold.

*Hint:*
- *To accomplish this, you will first need to group the data by state and month, then count the number of policies sold for each group. Afterwards, you will need to sort the data by the count of policies sold in descending order.*
- *Next, you will select the top 3 states with the highest number of policies sold.*
- *Finally, you will create a new DataFrame that contains the number of policies sold by month for each of the top 3 states.*

In [56]:
# Identify the top 3 states with the highest total policies sold
top_states = policies_by_state_month.groupby("State")["Policy"].sum().nlargest(3).index

# Filter the dataset to include only the top 3 states
filtered_policies = policies_by_state_month[policies_by_state_month["State"].isin(top_states)]

# Pivot the data to format it as a table (States as rows, Months as columns)
top_states_table = filtered_policies.pivot(index="State", columns="Month", values="Policy")

# Display the new DataFrame
print(top_states_table)


Month          1     2
State                 
Arizona     1008   929
California  1918  1634
Oregon      1565  1344


7. The marketing team wants to analyze the effect of different marketing channels on the customer response rate.

Hint: You can use melt to unpivot the data and create a table that shows the customer response rate (those who responded "Yes") by marketing channel.

External Resources for Data Filtering: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

In [61]:
# Select relevant marketing channel columns
channels = ["Sales Channel", "Response"]

# Melt the DataFrame to unpivot marketing channels
df_melted = df[channels].melt(id_vars=["Response"], var_name="Marketing_Channel", value_name="Channel_Type")#- .melt(id_vars=["Response"]) → Keeps the "Response" column unchanged while converting other columns into a long-format structure.

# Filter for customers who responded "Yes"
df_positive_response = df_melted[df_melted["Response"] == "Yes"]

# Calculate response rate for each marketing channel
response_rates = df_positive_response.groupby("Channel_Type")["Response"].count() / df_melted.groupby("Channel_Type")["Response"].count()

# Convert to DataFrame
response_rates_df = response_rates.reset_index().rename(columns={"Response": "Response Rate"})


response_rates_df


Channel_Type  Response Rate
0        Agent       0.190746
1       Branch       0.113787
2  Call Center       0.109786
3          Web       0.117141